### ISSUE TITLE
Raise an exception for non-prime p in FiniteFIeld(p)

### ISSUE THREAD

**Issue by normalhuman** [sympy/sympy#13886]
This is meant to be someone's first issue to fix. As noted in #9544, the FiniteField class does not check that the input  makes sense. 
```
>>> FiniteField(42)
GF(42)    # there is no such thing
>>> FiniteField(9).characteristic()
9   # GF(9) exists but it is of characteristic 3; not this object. 
```
Only prime orders are currently by this class, so its `__init__` method [here](https://github.com/sympy/sympy/blob/9670ff89dd8bff7878c6d636ec4f493b87d092df/sympy/polys/domains/finitefield.py#L28) should check the primality of the argument and raise `ValueError("Only fields of prime order are supported")` if it isn't. SymPy has [isprime](http://docs.sympy.org/latest/modules/ntheory.html#sympy.ntheory.primetest.isprime) for primality testing. 

A test should be added as well: `raises(ValueError, lambda: FiniteField(42))`. Somewhere [here](https://github.com/sympy/sympy/tree/9670ff89dd8bff7878c6d636ec4f493b87d092df/sympy/polys/domains/tests), perhaps in a new file? There is no `test_finitefield.py` yet.

**Comment by samarthsinghal1** [comment]
I would like to work on this issue.

**Comment by samarthsinghal1** [comment]
#13893 - FF(n) function is being called in test_densearith.py with non prime n and the test is failing as it is expecting an output. How should I go about fixing it? Should I include it in a try-except block to handle that ValueError?

**Comment by normalhuman** [comment]
No, don't put try-except block. This needs further investigation: why does the test want a FIniteField of nonprime order, what does it expect to do with it?

**Comment by samarthsinghal1** [comment]
test_dup_mul() which tests the multiplication of dense polynomials in '' K[x] '' calls the following -

K = FF(6)
assert dup_mul([K(2),K(1)],[K(3),K(4)],K) == [K(5), K(4)]

test_dmp_mul() is also calling using it in a similar way.

**Comment by normalhuman** [comment]
Looking into this, it appears the class is used in a way that ignores its FiniteField name. It is used mostly to represent the ring Z/(nZ) which makes sense for any n. The name and notation are wrong, and "characteristic" doesn't seem to be the right term to use, but trying to change this looks like fixing the wrong problem. 

I'm going to close this issue because raising an exception isn't a way to deal with this situation; the whole thing needs to be ripped apart and a proper FiniteField class put in its place. Sorry about that.

**Comment by jksuom** [comment]
One way to deal with this is to define another class `FiniteRing` that would do what `FiniteField` does now. Then the invocations of `FiniteField` with non-prime (or non-primepower) arguments should be deprecated and passed on to `FiniteRing`.

**Comment by samarthsinghal1** [comment]
I can define the class and look for all the invocations of FF(n) with non-prime n and replace the calls with the FiniteRing. But the whole class would be the same as FiniteField except the primality testing?

**Comment by jksuom** [comment]
`FiniteRing` could be the same as `FiniteField` is now, but `FiniteField` should (eventually) be fixed to create a field for prime power arguments `p^n`  instead of the ring `Z/p^nZ`.

**Comment by skirpichev** [comment]
Looking on the history, starting from 31c9aa43ff, it seems that author intention was that ``FF(n)`` should be a finite *field*, not ring.

Apparently, not everything (i.e. the characteristic of the field is wrong) is working as expected for non-prime Galois fields, i.e. if order is a perfect power.  So, initial approach to address this problem does make sense.  Few broken tests should be replaced, using fields with prime orders.  For perfect powers you could raise NotImplementedError, and ValueError for other non-prime values of ``mod``.

**Comment by gschintgen** [comment]
This is not fixed and current behavior is highly counter-intuitive/misleading. I'm reopening the issue.

**Comment by oscarbenjamin** [comment]
Is there an implementation of *proper* non-prime finite fields anywhere e.g. `GF(4)`?

**Comment by gschintgen** [comment]
> 
> 
> Is there an implementation of _proper_ non-prime finite fields anywhere e.g. `GF(4)`?

Maybe this:
```
In [41]: Z2x = FiniteField(2).old_poly_ring(x)
In [42]: GF4 = Z2x.quotient_ring([x**2+x+1])
In [43]: GF4
Out[43]: GF(2)[x]/<x**2 + x + 1>
```
(I don't know the history/significance of `old_poly_ring`.)

**Comment by gschintgen** [comment]
> (I don't know the history/significance of `old_poly_ring`.)

From https://github.com/sympy/sympy/pull/2126 :
> switched PolynomialRing and FractionField to use PolyElement and FracElement instead of DMP and DMF
> There is one downside of this PR. I couldn't, yet, remove old PolynomialRing and FractionField (those using DMP and DMF), because I didn't want to dig into agca module

**Comment by jksuom** [comment]
> Is there an implementation of proper non-prime finite fields anywhere e.g. GF(4)?

https://github.com/sympy/sympy/blob/4a0e1f003c766f5a7ea400bee2e37db53f7cb6ba/sympy/polys/agca/tests/test_extensions.py#L18-L28

**Comment by gschintgen** [comment]
Thanks @jksuom! I think that should definitely be part of the public documentation.

**Comment by gschintgen** [comment]
So this could serve as the basis to create a proper GF domain?
Inverse / division is missing, but that shouldn't be too hard to add. At least from a theoretical point of view it's fairly easy (`a**(-1) == a**(p**n-2)` or Euclidean algorithm). And then there's the question of the irreducible polynomial.

**Comment by jksuom** [comment]
> this could serve as the basis to create a proper GF domain?

That was my idea as well. I also had a version of division implemented: https://github.com/sympy/sympy/pull/19593.
